source/common/http/conn_manager_impl.cc

```diff
@@ -775,15 +775,36 @@ void ConnectionManagerImpl::ActiveStream::decodeData(ActiveStreamDecoderFilter*
 
   for (; entry != decoder_filters_.end(); entry++) {
     ASSERT(!(state_.filter_call_state_ & FilterCallState::DecodeData));
+
+    bool end_stream_no_trailers = end_stream && !request_trailers_;
+    if (end_stream_no_trailers) {
+      state_.filter_call_state_ |= FilterCallState::LastDataFrame;
+    }
     state_.filter_call_state_ |= FilterCallState::DecodeData;
-    FilterDataStatus status = (*entry)->handle_->decodeData(data, end_stream);
+    FilterDataStatus status = (*entry)->handle_->decodeData(data, end_stream_no_trailers);
     state_.filter_call_state_ &= ~FilterCallState::DecodeData;
+    if (end_stream_no_trailers) {
+      state_.filter_call_state_ &= ~FilterCallState::LastDataFrame;
+    }
     ENVOY_STREAM_LOG(trace, "decode data called: filter={} status={}", *this,
                      static_cast<const void*>((*entry).get()), static_cast<uint64_t>(status));
     if (!(*entry)->commonHandleAfterDataCallback(status, data, state_.decoder_filters_streaming_)) {
       return;
     }
   }
+
+  if (end_stream && request_trailers_) {
+    decodeTrailers(filter, *request_trailers_);
+  }
+}
+
+void ConnectionManagerImpl::ActiveStream::addDecodedTrailers(HeaderMapPtr&& trailers) {
+  if (state_.filter_call_state_ & FilterCallState::LastDataFrame) {
+    request_trailers_ = std::move(trailers);
```

Comment:
If I'm correct above that we can only set request_trailers_once now, can we also have an invariant ASSERT that they're empty when this is called?